geode-protobuf/src/test/java/org/apache/geode/protocol/RoundTripCacheConnectionJUnitTest.java

```diff
@@ -291,56 +298,101 @@ public void testConnectionCountIsProperlyDecremented() throws Exception {
 
     CacheServer cacheServer = cache.addCacheServer();
     final int cacheServerPort = AvailablePortHelper.getRandomAvailableTCPPort();
     cacheServer.setPort(cacheServerPort);
-    cacheServer.setMaxConnections(16);
-    cacheServer.setMaxThreads(16);
+    cacheServer.setMaxConnections(connections);
+    cacheServer.setMaxThreads(threads);
     cacheServer.start();
 
     AcceptorImpl acceptor = ((CacheServerImpl) cacheServer).getAcceptor();
 
-    // Start 16 sockets, which is exactly the maximum that the server will support.
-    Socket[] sockets = new Socket[16];
-    for (int i = 0; i < 16; i++) {
-      Socket socket = new Socket("localhost", cacheServerPort);
-      sockets[i] = socket;
-      Awaitility.await().atMost(5, TimeUnit.SECONDS).until(socket::isConnected);
-      socket.getOutputStream()
-          .write(CommunicationMode.ProtobufClientServerProtocol.getModeNumber());
+    if (isSelector) {
+      assertTrue(acceptor.isSelector());
+    } else {
+      assertFalse(acceptor.isSelector());
     }
 
-    // try to start a new socket, expecting it to be disconnected.
-    try (Socket socket = new Socket("localhost", cacheServerPort)) {
-      Awaitility.await().atMost(5, TimeUnit.SECONDS).until(socket::isConnected);
-      socket.getOutputStream()
-          .write(CommunicationMode.ProtobufClientServerProtocol.getModeNumber());
-      assertEquals(-1, socket.getInputStream().read()); // EOF implies disconnected.
-    }
+    validateSocketCreationAndDestruction(cacheServerPort, connections);
 
-    for (Socket currentSocket : sockets) {
-      currentSocket.close();
-    }
+    // Once all connections are closed, the acceptor should have a connection count of 0.
+    Awaitility.await().atMost(5, TimeUnit.SECONDS)
+        .until(() -> acceptor.getClientServerCnxCount() == 0);
+
+    // do it again, just to be sure there's no leak somewhere else.
+    validateSocketCreationAndDestruction(cacheServerPort, connections);
 
     // Once all connections are closed, the acceptor should have a connection count of 0.
     Awaitility.await().atMost(5, TimeUnit.SECONDS)
         .until(() -> acceptor.getClientServerCnxCount() == 0);
 
-    // Try to start 16 new connections, again at the limit.
-    for (int i = 0; i < 16; i++) {
-      Socket socket = new Socket("localhost", cacheServerPort);
-      sockets[i] = socket;
+  }
+
+  // Start exactly as many that the server will support, check that they work.
+  // test that creating one more causes it to be disconnected.
+  // Close all the sockets when we're done.
+  private void validateSocketCreationAndDestruction(int cacheServerPort, int connections)
+      throws Exception {
+    final Socket[] sockets = new Socket[connections];
+
+    ExecutorService executor = Executors.newFixedThreadPool(20);
+
+    // Used to assert the exception is non-null.
+    ArrayList<Callable<Exception>> callables = new ArrayList<>();
+
+    for (int i = 0; i < connections; i++) {
```

Comment:
could this be a `map().collect()`

Reply:
Nvm. getting the index in there wouldn't be as pretty in Java as one would hope.